cocotb/regression.py

```diff
@@ -641,7 +641,9 @@ class TestFactory:
     """
 
     def __init__(self, test_function, *args, **kwargs):
-        if not isinstance(test_function, cocotb.coroutine):
+        if inspect.isasyncgenfunction(test_function):
+            raise TypeError("Found yield in async function '%s'" % test_function.__name__)
+        if not (isinstance(test_function, cocotb.coroutine) or inspect.iscoroutinefunction(test_function)):
             raise TypeError("TestFactory requires a cocotb coroutine")
         self.test_function = test_function
         self.name = self.test_function._func.__qualname__
```

Comment:
I think the best fix would be:
```suggestion
        self.name = self.test_function.__qualname__
```
And add the `__qualname__` attribute to `coroutime` in the constructor.